dev/com.ibm.ws.transport.http/src/com/ibm/ws/netty/upgrade/NettyServletUpgradeHandler.java

```diff
@@ -105,28 +246,60 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
     }
-    
+
     public void immediateTimeout() {
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
             Tr.debug(this, tc, "Inside immediate timeout! " + channel);
         }
+        if(isReadingAsync) {
+            cancelTimer();
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(this, tc, "Timeout hit for channel on async" + channel);
+            }
+            StringBuilder error = new StringBuilder();
+            error.append("Socket operation timed out before it could be completed local=");
+            error.append(readContext.getInterface().getLocalAddress().getHostName()).append("/");
+            error.append(readContext.getInterface().getLocalAddress().getHostAddress()).append(":");
+            error.append(readContext.getInterface().getLocalPort());
+            error.append(" remote=");
+            error.append(readContext.getInterface().getRemoteAddress().getHostName()).append("/");
+            error.append(readContext.getInterface().getRemoteAddress().getHostAddress()).append(":");
+            error.append(readContext.getInterface().getRemotePort());
+
+            HttpDispatcher.getExecutorService().execute(() -> {
+                try {
+                    getReadListener().error(vc, readContext, new SocketTimeoutException(error.toString()));
+                } catch (Exception e) {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(this, tc,
+                                "NettyServletUpgradeHandler immediateTimeout hit Exception on error callback: " + e);
+                    }
+                }
+            });
+            isReadingAsync = false;
+            return;
+        }
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(this, tc, "NettyServletUpgradeHandler immediateTimeout NOT found async!!");
+        }
         immediateTimeout.getAndSet(true);
         signalReadReady();
-        // TODO Loop to make sure no others reads are taking place
-        while(readLock.hasQueuedThreads()) { // Queue here until no queued threads are available
+        while (waitingThreads.get() > 0) { // Queue here until no queued threads are available
             if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                 Tr.debug(this, tc, "NettyServletUpgradeHandler immediateTimeout waiting to finish immediate timeout on channel: " + channel);
             }
-            // TODO If this is kept and not removed when disabling auto read, switch this logic to provide assertion of not running
+            // If this is kept and not removed when disabling auto read, switch this logic to provide assertion of not running
             // in the event loop and run a repeatable task with ScheduledExecutorService
             try {
-                Thread.sleep(1000);
+                Thread.sleep(500);
```

Comment:
Can we verify if we need this sleep?

Reply:
I do think we need it for synchronous immediate read but with the async read changes I don't know if there is anything using this immediate call. I opened https://github.com/OpenLiberty/open-liberty/issues/32229 to follow up on this